KT-51297: Allow suspend from any thread from ObjC

ahmedre opened the pull request:
This patch adds a binary compiler argument to allow not crashing if a
suspend function is called from a non-Main dispatcher from ObjC or
Swift. This is relevant for people using the new memory model, in which
this restriction may be relaxed.

ahmedre commented:
@SvyatoslavScherbina this is per your comment [here](https://youtrack.jetbrains.com/issue/KT-51297/Native-allow-calling-Kotlin-suspend-functions-on-non-main-thread).

nikolaykasyanov commented on kotlin-native/backend.native/tests/objcexport/coroutines.swift:
```diff
@@ -48,6 +48,36 @@ private func testUnitCallSimple() throws {
+#endif
+#endif
+
+    try assertNil(error)
```
You'd need to ensure that the background queue has finished executing blocks by the time this is asserted.

ahmedre replied:
done.

SvyatoslavScherbina commented on kotlin-native/runtime/src/main/cpp/ObjCExportCoroutines.mm:
```diff
@@ -36,8 +36,8 @@
 
-extern "C" OBJ_GETTER(Kotlin_ObjCExport_createContinuationArgument, id completion, const TypeInfo** exceptionTypes) {
-  if (pthread_main_np() != 1) {
+extern "C" OBJ_GETTER(Kotlin_ObjCExport_createContinuationArgument, KBoolean allowAnyThread, id completion, const TypeInfo** exceptionTypes) {
```
There is a significantly simpler way to pass the value from compiler arguments to runtime.
See e.g. https://github.com/JetBrains/kotlin/blob/19658b3a4706284eeb21480f53fe2740ba60ce88/kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt#L2779 and https://github.com/JetBrains/kotlin/blob/3798920183e9180795726ac89d18c4355dac4520/kotlin-native/runtime/src/main/cpp/CompilerConstants.cpp#L17

Please use this approach.

ahmedre replied:
done

ahmedre commented:
changes:
- renamed the flag to something hopefully a little more reasonable (`allowLaunchSuspendFunctionFromThread`)
- renamed parameters to `main` and `any`
- fixed the test to properly wait for the work to finish

SvyatoslavScherbina commented on kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/AllowLaunchSuspendFunctionFromThread.kt:
```diff
@@ -0,0 +1,23 @@
+
+package org.jetbrains.kotlin.backend.konan
+
+enum class AllowLaunchSuspendFunctionFromThread {
```
The name and the descriptions below are a bit misleading: this flag covers only the case when calling a Kotlin `suspend` function from Swift or Objective-C.
When calling from Kotlin, there are no such limitations, and the flag does not affect this.

Same for other similar names. Same for the commit message.

ahmedre replied:
changed - happy to change this if you may recommend a better suggestion!

SvyatoslavScherbina replied:
Thanks!

The new version is still not entirely correct -- even on Apple targets, there were no thread restrictions when calling `suspend` function from Kotlin.
It is not the target that matters with the existing restriction, but the fact that `suspend` function is called from Objective-C (or Swift).

What do you think about something like `ObjCExportSuspendFunctionLaunchThreadRestriction`?

ahmedre replied:
done.

ahmedre replied:
sounds good, will do!

ahmedre replied:
done

SvyatoslavScherbina commented on kotlin-native/backend.native/tests/objcexport/coroutines.swift:
```diff
@@ -48,6 +48,39 @@ private func testUnitCallSimple() throws {
+#endif
+
+    group.wait()
+    try assertNil(error)
```
```suggestion
    try assertEquals(actual: completionCalled, expected: 1)
    try assertNil(error)
```

ahmedre replied:
done, had to also update the test to work in the else case of the ifdef where the flag isn't passed.

ahmedre commented on kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/ObjCExportSuspendFunctionLaunchThreadRestriction.kt:
```diff
@@ -0,0 +1,23 @@
+    ;
+
+    companion object {
+        val DEFAULT = MAIN
```
yes - will remove.

ahmedre replied:
done

SvyatoslavScherbina reviewed (approved):
Thanks!
